File: app/src/main/java/com/two_two/checkreaction/startscreen/StartScreenPresenter.java
package com.two_two.checkreaction.startscreen;

import com.two_two.checkreaction.model.App;
import com.two_two.checkreaction.model.game.TestType;

import java.util.ArrayList;

/**
 * Created by Dmitry Borodin on 1/3/2016.
 */
public class StartScreenPresenter {

    private static volatile StartScreenPresenter sInstance;
    private ArrayList<StartActivityContract> mCurrentActivity = new ArrayList<>(2);

    public static StartScreenPresenter getInstance() {
        StartScreenPresenter instance = sInstance;
        if (instance == null) {
            synchronized (StartScreenPresenter.class) {
                instance = sInstance;
                if (instance == null) {
                    sInstance = instance = new StartScreenPresenter();
                }
            }
        }
        return instance;
    }

    private StartScreenPresenter() {
    }

    // Section: Methods for view

    public void registerActivity(StartActivityContract activity) {
        mCurrentActivity.add(activity);
    }

    public void unregisterActivity(StartActivityContract activity) {
        mCurrentActivity.remove(activity);
    }

    public void startSimpleTest() {
        notifyStartTest(TestType.SIMPLE_TEST);
    }

    public void startComplexTest() {
        notifyStartTest(TestType.COMPLEX_TEST);
    }

    /**
     * If username is valid, this will be used in global score results in the cloud.
     * If not valid, nothing happens, previous or default name will remain.
     *
     * @param username name to save
     */
    public void saveUsername(String username) {
        if (isUsernameCurrect(username)) App.getInstance().getLocalData().setUsername(username);
    }

    private boolean isUsernameCurrect(String username) {
        if (username.isEmpty() || username.length() > 100) return false;
        return true;
    }

    private void notifyStartTest(TestType testType) {
        for (StartActivityContract activity : mCurrentActivity) {
            activity.startTest(testType);
        }
    }
}

File: app/src/main/java/com/two_two/checkreaction/ui/startscreen/StartScreenPresenter.kt
package com.two_two.checkreaction.ui.startscreen

import com.two_two.checkreaction.models.App
import com.two_two.checkreaction.models.game.TestType
import com.two_two.checkreaction.utils.Constants

/**
 * Created by Dmitry Borodin on 1/3/2016.
 */
class StartScreenPresenter() {

    private var mCurrentActivity: StartActivityContract? = null

    fun bindActivity(activity: StartActivityContract) {
        mCurrentActivity = activity
    }

    fun unbindActivity() {
        mCurrentActivity = null
    }

    fun startSimpleTest() {
        startTest(TestType.SIMPLE_TEST)
    }

    fun startComplexTest() {
        startTest(TestType.COMPLEX_TEST)
    }

    fun startScienceTest() {
        startTest(TestType.SCIENCE_TEST)
    }

    /**
     * If username is valid, this will be used in global score results in the cloud.
     * If not valid, nothing happens, previous or default name will remain.

     * @param username name to save
     */
    fun saveUsername(username: String) {
        if (isUsernameCorrect(username)) App.instance.localData?.username = username
    }

    private fun isUsernameCorrect(username: String): Boolean {
        return !(username.isEmpty() || username.length > Constants.USERNAME_LEIGHT_LIMIT)
    }

    private fun startTest(testType: TestType) {
        mCurrentActivity?.startTest(testType)
    }
}
